projects/igniteui-angular/src/lib/grids/grid-mrl-navigation.service.ts

```diff
@@ -0,0 +1,466 @@
+            });
+            super.navigateUp(currentRowEl, rowIndex, lastVisibleIndex, cell);
+        }
+    }
+
+    private focusCellUpFromLayout(cell, isSummary = false) {
+        const columnLayout = cell.column.parent;
+        const element = cell.nativeElement.parentElement;
+
+        const currentRowStart = cell.rowStart;
+        const currentColStart = cell.colStart;
+
+        // element up is from the same layout
+        let upperElementColumn = columnLayout.children.find(c =>
+            (c.rowEnd === currentRowStart || c.rowStart + c.gridRowSpan === currentRowStart)  &&
+            c.colStart <= currentColStart &&
+            (currentColStart < c.colEnd || currentColStart < c.colStart + c.gridColumnSpan));
+
+        const columnIndex = columnLayout.children.toArray().indexOf(upperElementColumn);
+        const upperElement = element.children[columnIndex];
+
+        if (!upperElement) {
+            const layoutRowEnd = this.grid.multiRowLayoutRowSize + 1;
+            upperElementColumn = columnLayout.children.find(c =>
+                (c.rowEnd === layoutRowEnd || c.rowStart + c.gridRowSpan === layoutRowEnd) &&
+                c.colStart <= currentColStart &&
+                (currentColStart < c.colEnd || currentColStart < c.colStart + c.gridColumnSpan));
+
+            const prevIndex = cell.row.index - 1;
+            let prevRow;
+            const rowElement = cell.row.nativeElement;
+            const containerTopOffset = parseInt(this.verticalDisplayContainerElement.style.top, 10);
+            if (prevIndex >= 0 && (!rowElement.previousElementSibling ||
+                rowElement.previousElementSibling.offsetTop < Math.abs(containerTopOffset))) {
+                this.grid.nativeElement.focus({ preventScroll: true });
+                this.grid.verticalScrollContainer.onChunkLoad
+                    .pipe(first())
+                    .subscribe(() => {
+                        prevRow = this.grid.getRowByIndex(prevIndex);
+                        if (prevRow && prevRow.cells) {
+                            this._focusCell(upperElementColumn.cells.find((c) => c.rowIndex === prevRow.index).nativeElement);
+                        } else if (prevRow) {
+                            prevRow.nativeElement.focus({ preventScroll: true });
+                        }
+                    });
+                this.grid.verticalScrollContainer.scrollTo(prevIndex);
+            } else {
+                prevRow = this.grid.getRowByIndex(prevIndex);
+                if (prevRow && prevRow.cells) {
+                    this._focusCell(upperElementColumn.cells.find((c) => c.rowIndex === prevRow.index).nativeElement);
+                } else if (prevRow) {
+                    prevRow.nativeElement.focus({ preventScroll: true });
+                }
+            }
+            return;
+        }
+        this._focusCell(upperElement);
+    }
+
+    private focusCellDownFromLayout(cell, isSummary = false) {
```

Comment:
all four functions (down, up, left, right) look similar. I am pretty sure locating the next cell can be done with more granular functions and the overarching flow can then be put into a single one saving code. This should be true at least for the Up/Down Next/Prev pairs.